rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,674 @@
+
+   Multiple `requires` and `ensures` clauses are allowed on the same function,
+   they are implicitly logically conjoined.
+
+
+2. Next, Kani must make sure that the approximation we specified actually holds
+   by **checking** the contract against the implementation. This is in contrast
+   to ["stubbing"][stubbing], where the approximation is blindly trusted.
+
+   The contract must always overapproximate the function to guarantee soundness.
+   To facilitate the check Kani needs a suitable environment to verify the
+   function in. For this proposal the environment must be provided by us (the
+   users). See [future possibilities](#future-possibilities) for a discussion
+   about the arising soundness issues and their remedies.
+
+   We provide the checking environment for our contract with a special new
+   `proof_for_contract` harness.
+
+   ```rs
+   #[kani::proof_for_contract(my_div)]
+   fn my_div_harness() {
+     my_div(kani::any(), kani::any())
+   }
+   ```
+
+   Similar to a verification harness for any other function, we are supposed to
+   create all possible input combinations the function can encounter, then call
+   the function at least once with those abstract inputs. If we forget to call
+   `my_div` Kani reports an error.
+   
+   Unlike a unit-test we can however elide any checks of the output and
+   post-call state. Instead Kani uses the conditions we specified in the
+   contract as checks. Preconditions (`requires`) are inserted as `kani::assume`
+   *before* the call to `my_div`, limiting the generic inputs to those the
+   function is actually defined for. Postconditions (`ensures`) are inserted as
+   `kani::assert` checks *after* the call to `my_div`.
+
+   The expanded version of our harness and function is equivalent to the following:
+
+   ```rs
+   #[kani::proof]
+   fn my_div_harness() {
+     let dividend = kani::any();
+     let divisor = kani::any();
+     kani::assume(divisor != 0); // requires
+     let result = my_div(dividend, divisor);
+     kani::assert(result <= dividend); // ensures
+   }
+   ```
+
+   This expanded harness is then verified like any other harness but also gives
+   the green light for the next step: verified stubbing.
+
+3. In the last phase the **verified** contract is ready for us to use to
+   **stub** other harnesses.
+
+   Unlike in regular stubbing, there has to be at least one associated
+   `proof_for_contract` harness for each function to stub *and* it requires all
+   such harnesses to pass verification before attempting verification of any
+   harnesses that use it as a stub.
+
+   A possible harness that uses our `my_div` contract could be the following:
+
+   ```rs
+   #[kani::proof]
+   #[kani::stub_verified(my_div)]
+   fn use_div() {
+     let v = vec![...];
+     let some_idx = my_div(v.len() - 1, 3);
+     v[some_idx];
+   }
+   ```
+
+   To use the contract as a stub Kani must first ensure the calling context is
```

Comment:
Can you use a more direct explanation like "at a call site where the contract is used as abstraction for the call, preconditions are asserted, and a nondet value satisfying the post condition is produced"

Reply:
Passive voice 😛